### PR TITLE
Add CI multicluster + OIDC environment

### LINKED ISSUE #2376: CI doesn't test deployment to other clusters so it breaks.
### Description:

The current CI for Kubeapps does not test deploying to a cluster other than the one on which Kubeapps is installed, so we occasionally get breakage and only find out when trying to create a demo :P.

### PR DESCRIPTION
**Pull request by antgamdia**
### Description of the change

(note: this PR supersedes #2610. It's better to split the multicluster environment and the general improvements)

This PR updates our integration tests to use the multicluster environment so that we can detect issues and errors (as proposed in #2376).

The environment is based on the one we already use for dev purposes, that is: dex/openldap + two kind clusters.
There are, also, a couple of tweaks since the integration tests run inside a container; therefore, the ingress URL is not available and there are some issues related to the cookies. 

### Benefits

Testing Kubeapps in multicluster environments with OIDC.

### Possible drawbacks

As the complexity of the CI increases, the flakiness of the overall CI systems might also increase. 

### Applicable issues

  - fixes #2376

### Additional information

This PR depends upon #2618  to be merged.

As an improvement, we can also consider including the OIDC via Pinniped. It shouldn't be a big deal since all the multicluster effort is already done in this PR.

### CODE REVIEW

File: integration/use-cases/add-multicluster-deployment.js
Comment (andresmgot) [inline integration/use-cases/add-multicluster-deployment.js:8]: I would not add this `process.env.ADMIN_TOKEN` (just to make clear this test won't work with a token)

File: integration/use-cases/create-private-registry.js
Comment (andresmgot) [inline integration/use-cases/create-private-registry.js:21]: unless there is a reason that waiting for the button doesn't work I won't add these kind of waits, it just makes the tests more difficult to maintain.

File: integration/use-cases/create-private-registry.js
Comment (andresmgot) [inline integration/use-cases/create-private-registry.js:50]: we can remove this try-catch now

File: integration/use-cases/create-private-registry.js
Comment (andresmgot) [inline integration/use-cases/create-private-registry.js:67]: same, we can remove the try-catch

File: integration/use-cases/missing-permissions.js
Comment (andresmgot) [inline integration/use-cases/missing-permissions.js:19]: this better waits for the text "apache" to appear

File: integration/use-cases/missing-permissions.js
Comment (andresmgot) [inline integration/use-cases/missing-permissions.js:31]: this should not be needed (you may want to increase the timeout below)

File: integration/use-cases/operator-deployment.js
Comment (andresmgot) [inline integration/use-cases/operator-deployment.js:17]: same, wait for "prometheus" (not sure if the timeout below works).

File: integration/use-cases/operator-deployment.js
Comment (andresmgot) [inline integration/use-cases/operator-deployment.js:41]: not sure if you already replied, why the `isAlreadyDeployed`?

File: integration/use-cases/operator-deployment.js
Comment (andresmgot) [inline integration/use-cases/operator-deployment.js:50]: the only `timeout` option working is for the `toMatch` function:

https://github.com/smooth-code/jest-puppeteer/blob/master/packages/expect-puppeteer/README.md#expectinstancetomatchmatcher-options

File: script/e2e-test.sh
Comment (andresmgot) [inline script/e2e-test.sh:352]: you can just use a single line for this

File: script/e2e-test.sh
Comment (andresmgot) [inline script/e2e-test.sh:371]: and the admin user?

**Review by andresmgot — COMMENTED** [review]
Thanks @antgamdia. This is looking good, just some minor comments.

File: integration/use-cases/create-private-registry.js
Comment (antgamdia) [reply integration/use-cases/create-private-registry.js:50]: When testing in my fork I didn't build the new CI images (for saving time), and I would still need these lines. Now they can be safely removed, thanks for the reminder!

File: integration/use-cases/operator-deployment.js
Comment (antgamdia) [reply integration/use-cases/operator-deployment.js:41]: Yep (sorry for dropping the former PR, but I wanted to split the changes for ease the review process)
Check it at https://github.com/kubeapps/kubeapps/pull/2610#discussion_r602470780
Since uninstalling the operator is not that easy, I left this `if` for those cases where we are executing the tests locally. I don't mind removing it if you want, though.

File: script/e2e-test.sh
Comment (antgamdia) [reply script/e2e-test.sh:371]: It is already created with cluster-admin role with the default RBAC (`kubeapps-operator@example.com`). 
This step is just modifying the `kubeapps-user@example.com`, but we can perform this step in the same place as the operator user for cohesion if you may prefer.

File: integration/use-cases/operator-deployment.js
Comment (andresmgot) [reply integration/use-cases/operator-deployment.js]: it should be fine (if for some reason this is bypassing the operator installation, it should fail when creating an instance).

File: script/e2e-test.sh
Comment (andresmgot) [reply script/e2e-test.sh:372]: I see, yes, better to have them in the same place.

**Review by andresmgot — APPROVED** [review]
LGTM when you fix the current error.

File: integration/jest.sequencer.js
Comment (antgamdia) [inline integration/jest.sequencer.js:6]: Despite the tests use to be executed in alphabetical order, there is no actual guarantee of that. It was causing a failure in the multicluster test because the secret created in the secret-repo test wasn't being found.
This sequencer just ensures this alphabetical order (I've pushed the integration img with this dep as well)

File: script/libtest.sh
Comment (antgamdia) [inline script/libtest.sh:44]: These changes are not strictly necessary (the main error was I didn't pass the db password :facepalm: ), but I think they make the function work as expected, I mean, performing an actual waiting, rather than digesting the exit_code for logging purposes.

**Comment by antgamdia** [comment]
> LGTM when you fix the current error.

Despite my sorrow and regardless of how I want much to press the merge button :P, I would need another +1 just to make sure the new changes are ok with you as well.

**Comment by absoludity** [comment]
> > LGTM when you fix the current error.
> 
> Despite my sorrow and regardless of how I want much to press the merge button :P, I would need another +1 just to make sure the new changes are ok with you as well.

Hah! Andres can merge it if happy with the changes (I thought it might have just been a few, but I see lots of commits, so I'll let Andres continue it). Enjoy your break o/

File: script/e2e-test.sh
Comment (andresmgot) [inline script/e2e-test.sh]: I am not sure you should need these?

The clusterrole/clusterrolebinding should be adopted by the new version and the apprepository should not exist because we are using `--set apprepository.initialRepos=null`

File: script/libtest.sh
Comment (andresmgot) [inline script/libtest.sh:52]: please don't add sleeps this long. I don't think we need to retry here.

**Review by andresmgot — COMMENTED** [review]
Thanks @antgamdia. I have some (minor) comments

File: integration/jest.sequencer.js
Comment (absoludity) [reply integration/jest.sequencer.js:6]: We don't yet support app repos (let alone secret ones) on other clusters (#1982)  because the interface to sync with the asset svc is via the DB (#2394) (left over from the read-only monocular).

Comment (antgamdia) [reply integration/jest.sequencer.js:6]: Thanks for the clarification, I always forget about this issue. 
Perhaps we should name the tests like "01-xxxx.js", "00-xxx.js" just to make clear that they are being executed in order.

File: integration/use-cases/create-private-registry.js
Comment (antgamdia) [reply integration/use-cases/create-private-registry.js:48]: No, I remove them indeed. However, the CI started failing since the selectors weren't been found, so I reverted the change.

https://app.circleci.com/pipelines/github/kubeapps/kubeapps/2614/workflows/dee98a52-3d20-47a6-9a76-a1fa16351eaf/jobs/42596/artifacts

```log
 FAIL  use-cases/operator-deployment.js (108.546 s)

 RUNS  ...
  ● Deploys an Operator

    TimeoutError: Element div.modal-dialog.modal-md > div > div.modal-body > div > div > cds-button:nth-child(2) (text: "Delete") not found

      waiting for function failed: timeout 30000ms exceeded
```

But, since the tests were failing because of other reasons, I'm gonna trigger again a new build and let's see what happens.
If it is still failing I guess it will be something related to the built image.

File: script/e2e-test.sh
Comment (antgamdia) [reply script/e2e-test.sh]: Unfortunately I think you saw a former commit, 7086d6f6d578c43d5c13b50400804d40a71da1f9 should fix it.
Thanks for the heads-up anyway :)

File: integration/jest.sequencer.js
Comment (andresmgot) [reply integration/jest.sequencer.js:6]: I know that private repositories are not supported in the second cluster but what I don't understand is why the order of the tests affects the result. I thought that creating a private repo in the default cluster caused an error when trying to deploy in the second cluster but that may not be it.

**Review by andresmgot — APPROVED** [review]
LGTM, I just don't understand very well why the tests need to be ordered.

File: integration/jest.sequencer.js
Comment (antgamdia) [reply integration/jest.sequencer.js:6]: Ok, ok, I see your point now and it may be an issue as you said:

When creating a private repo first, we are creating the docker credentials. Somehow, when deploying in the second cluster, it tries to retrieve these credentials (not available, though), and fails, therefore.

image

Let me replicate the issue in my dev environment and will file an issue with the details.

However, with regards to this PR, I think (for the sake of replicability) to still define an explicit order during the test suite execution.

Comment (antgamdia) [reply integration/jest.sequencer.js:6]: I wasn't able to reproduce this issue in my dev environment. I set up the multicluster env and: 1) tried to manually create the private repo and then a new deployment; 2) run the e2e test directly from `yarn start ...`.
In both cases, the credentials are being properly stored and the app deploys seamlessly as well.

That's odd... :S

Comment (andresmgot) [reply integration/jest.sequencer.js:6]: You probably added a namespaced private repository. If you create it in the `kubeapps` namespace, you are likely able to reproduce the issue.

**Pull request merged**
